File: go/141.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func hasCycle(head *ListNode) bool {
	x, y := head, head
	for x != nil && y != nil {
		if y.Next != nil {
			y = y.Next.Next
		} else {
			break
		}
		x = x.Next
		if x == y {
			return true
		}
	}
	return false
}

File: cpp/141.cpp

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
  bool hasCycle(ListNode *head) {
    ListNode *x{head}, *y{head};
    while (x != nullptr && y != nullptr) {
      if (y->next)
        y = y->next->next;
      else
        break;
      x = x->next;
      if (x == y)
        return true;
    }
    return false;
  }
};
